api/python/t4/bucket.py

```diff
@@ -0,0 +1,221 @@
+        """
+        # Ensure key ends in '/'.
+        if key[-1] != '/':
+            key = key + '/'
+
+        src_path = pathlib.Path(directory)
+        if not src_path.is_dir():
+            raise QuiltException("Provided directory does not exist")
+
+        files = src_path.rglob('*')
+        for f in files:
+            if not f.is_file():
+                continue
+            new_key = key + f.relative_to(src_path).as_posix()
+            new_path = self._uri + new_key
+            copy_file(f.resolve().as_uri(), new_path)
+
+    def keys(self):
+        """
+        Lists all keys in the bucket.
+
+        Returns:
+            list of strings
+        """
+        return list(map(lambda x: x.get('Key'), list_objects(self._bucket, '')))
+
+    def delete(self, key):
+        """
+        Deletes a key from the bucket.
+
+        Args:
+            key(str): key to delete
+
+        Returns:
+            None
+
+        Raises:
+            if delete fails
+        """
+        delete_object(self._bucket, key)
+
+    def fetch(self, key, path):
+        """
+        Fetches file (or files) at key to path.
+
+        If key ends in '/', then all files with the prefix key will match and will
+            be stored in a directory at path.
+        Otherwise, only one file will be fetched and it will be stored at path.
+
+        Args:
+            key(str): key in bucket to fetch
+            path(str): path in local filesystem to store file or files fetched
+
+        Returns:
+            None
+
+        Raises:
+            if path doesn't exist
+            if download fails
+        """
+        if not key.endswith('/'):
+            copy_file(self._uri + key, fix_url(path))
+            return
+
+        _download_dir(self._bucket, key, fix_url(path))
+
+    def get_meta(self, key):
+        """
+        Gets the metadata associated with a key in bucket.
+
+        Args:
+            key(str): key in bucket to get meta for
+
+        Returns:
+            dict of meta
+
+        Raises:
+            if download fails
+        """
+        response = s3_client.head_object(
+            Bucket=self._bucket,
+            Key=key
+        )
+        return response['Metadata']
+
+    def set_meta(self, key, meta):
+        """
+        Sets user metadata on key in bucket.
+
+        Args:
+            key(str): key in bucket to set meta for
+            meta(dict): value to set user metadata to
+
+        Returns:
+            None
+
+        Raises:
+            if put to bucket fails
+        """
+        meta = meta or {}
+        existing_meta = self.get_meta(key)
+        new_meta = existing_meta
+        new_meta['user_meta'] = json.dumps(meta)
+        s3_client.copy_object(
+            CopySource={
+                "Bucket": self._bucket,
+                "Key": key
+            },
+            Bucket=self._bucket,
+            Key=key,
+            MetadataDirective='REPLACE',
+            Metadata=new_meta
```

Comment:
Pretty sure this doesn't work - need the "helium" header. But really, just use the existing `copy_object` helper function.